### ISSUE TITLE
`pipe.pkl` is committed in the repo and contributors may use a  potentially stale or version-incompatible binary without realising it.

### ISSUE THREAD

**Issue by sh333299-sys** [Arnav-Singh-5080/IPL-Win-Probability-Prediction#375]
### Steps to Reproduce

**Bloat issue:**
1. Clone the repo: `git clone https://github.com/Arnav-Singh-5080/CricScope.git`
2. Run `git log --all --full-history -- pipe.pkl`
3. Observe that the binary is tracked in Git history

**Version mismatch issue:**
1. Clone the repo with a different scikit-learn version than what 
   was used to train the model
2. Run `streamlit run application.py`
3. Observe potential `ValueError` or `AttributeError` on model load

---

### Expected Behavior
- `pipe.pkl` should be listed in `.gitignore` so it is never committed
- A training script (e.g. `train_model.py`) should be provided so 
  contributors can regenerate `pipe.pkl` locally in one command
- The README should document the step:

```bash
# Step 1 — Generate the model locally before running the app
python train_model.py

# Step 2 — Run the app
streamlit run application.py
```

---

### Actual Behavior
`pipe.pkl` is committed in the repo and contributors may use a 
potentially stale or version-incompatible binary without realising it.

---

### Suggested Fix

**1. Add `pipe.pkl` to `.gitignore`:**
```gitignore
# ML model artifacts — regenerate locally using train_model.py
pipe.pkl
*.pkl
```

**2. Extract training logic into a standalone `train_model.py`:**
```python
import pandas as pd
import pickle
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.linear_model import LogisticRegression

# Load data
matches = pd.read_csv('matches.csv')
deliveries = pd.read_csv('deliveries.csv')

# --- feature engineering & pipeline definition here ---
# (move existing training code from application.py into this file)

# Save model
with open('pipe.pkl', 'wb') as f:
    pickle.dump(pipe, f)

print("✅ Model trained and saved to pipe.pkl")
```

**3. Update README setup steps:**
```markdown
## 4 — Train the model locally
python train_model.py

## 5 — Run the app
streamlit run application.py
```

**4. Remove `pipe.pkl` from Git history (optional but recommended):**
```bash
git rm --cached pipe.pkl
git commit -m "chore: remove pipe.pkl from version control"
```

---

### Files Affected
- `.gitignore` — add `pipe.pkl` / `*.pkl`
- `train_model.py` — new file to be created
- `README.md` — update setup instructions
- `pipe.pkl` — remove from tracking

---

### Labels
`bug` · `devops` · `good first issue` · `documentation`

---

### Additional Context
This is a standard ML engineering best practice. Separating training 
logic from app logic also makes the codebase easier to maintain — 
contributors can retrain the model with updated data (e.g. IPL 2021–2024) 
without modifying `application.py`.

This issue pairs well with the dataset update issue, since a 
`train_model.py` script would make retraining on new data trivial.

**Comment by aarushlohit** [comment]
Hi! I'd like to work on this issue under GSSoC 2026.

I reviewed the issue and agree that committing `pipe.pkl` directly to the repository can lead to version compatibility problems, stale model artifacts, repository bloat, and reduced reproducibility for contributors.

I can work on:

* Adding `pipe.pkl` (and other generated model artifacts where appropriate) to `.gitignore`
* Extracting the model training workflow into a dedicated `train_model.py` script
* Updating the application to clearly separate training and inference responsibilities
* Improving the README with reproducible setup and training instructions
* Verifying that contributors can regenerate the model locally with a documented workflow
* Reviewing whether additional generated artifacts should be excluded from version control

This will improve maintainability, reproducibility, and onboarding for future contributors while aligning the project with common ML engineering best practices.

Could you please assign this issue to me? Thank you!

@sh333299-sys
@Arnav-Singh-5080
